Issue #1746: Moving QUANTILE

hawkfish opened the pull request:
First pass at moving quantile implementation. There is more to do, but this has an important fix for the API by passing in `NULL` masks, which was a bug.

Commit:
Issue #1657: Windowed quantile

Move ResizeState into QuantileState.
Eventually this will store the index
for moving quantile instead of the
values (which will be passed in.)

Commit:
Merge branch 'hawkfish-temporal-structs' into hawkfish-moving-quantile

Commit:
Merge branch 'hawkfish-temporal-structs' into hawkfish-moving-quantile

Commit:
Issue #1680: Temporal struct types

Simplify quantile to to leverate
temporal struct types.

Commit:
Merge branch 'hawkfish-temporal-structs' into hawkfish-moving-quantile

Commit:
Merge branch 'hawkfish-moving-quantile' of https://github.com/hawkfish/duckdb into hawkfish-moving-quantile

hawkfish committed:
Merge pull request #58 from duckdb/master

Merge as of 2021-04-29

hawkfish committed:
Merge pull request #62 from duckdb/master

Merge as of 2021-05-03

Commit:
Issue #1657: Moving QUANTILE

Code and tests. Waiting on simpler MODE changes.

Commit:
Merge branch 'hawkfish-moving-quantile' of https://github.com/hawkfish/duckdb into hawkfish-moving-quantile

Commit:
Merge branch 'hawkfish-moving-mode' into hawkfish-moving-quantile

Commit:
Merge branch 'master' of github.com:duckdb/duckdb into hawkfish-moving-quantile

Commit:
Merge branch 'hawkfish-moving-quantile' of github.com:hawkfish/duckdb into hawkfish-moving-quantile

Commit:
Merge branch 'master' into hawkfish-moving-quantile

Commit:
Issue #1716: Moving MODE performance

Move reused objects into the class
to improve framed performance by 2x.

Commit:
Issue #1756: Moving QUANTILE

Rework the infrastructure for `Frame` function support.

Commit:
Merge branch 'master' into hawkfish-moving-quantile

Commit:
Merge branch 'hawkfish-moving-quantile' of https://github.com/hawkfish/duckdb into hawkfish-moving-quantile

Commit:
Issue #1756: Moving QUANTILE

Reduce frame data to a union of both frames.
Refactor QUANTILE to remove Value.

Commit:
Issue #1756: Moving QUANTILE

Fix std:min/max vs Win64.

Commit:
Issue #1756: Moving QUANTILE

Rename Frame to Window.
Add support for NULLs to Window API
Check for oversized frames.

Commit:
Merge branch 'master' into hawkfish-moving-quantile

Commit:
Issue #1716: Moving MODE

Add missing NULL detection and test it.

Commit:
Issue #1756: Moving QUANTILE

Reduce calls to Reset.
Removes one of the two main hot spots
in the fixed frame size test computation.

This code is shared with the segment tree
implementation, so it should help there too.

hawkfish commented:
The latest change is just a perf change, but it impacts all windowed aggregates by removing unnecessary memory allocations from most segment tree computations. 

The perf improvement for windowed quantile (10M rows, fixed frame) is about 20%:

```shell
./build/reldebug/test/unittest test/sql/window/test_moving_quantile.test_slow  7.42s user 0.30s system 99% cpu 7.728 total
./build/reldebug/test/unittest test/sql/window/test_moving_quantile.test_slow  6.10s user 0.26s system 96% cpu 6.582 total
```

Mytherin commented on test/sql/window/test_quantile_window.test_slow:
```diff
@@ -0,0 +1,21 @@
+require vector_size 512
+
+#statement ok
+#PRAGMA enable_verification
```
Is this disabled by accident?

hawkfish replied:
Yes. More precisely, I was timing queries and wanted to remove overhead. I'll restore it.

hawkfish replied:
After checking, I have removed it because it dramatically increases the run time.

hawkfish replied:
Upon review, this test file was an accident and should be removed. The other test file is more complete.

Mytherin commented on test/sql/window/test_moving_quantile.test_slow:
```diff
@@ -0,0 +1,41 @@
+#   Fixed window
+statement ok
+prepare fwm as
+select sum(m)
```
Can we check the result here?

hawkfish replied:
There are 10M values, but I will add a checksum.

Mytherin commented on test/sql/window/test_quantile_window.test_slow:
```diff
@@ -0,0 +1,21 @@
+create table rank100 as select b % 100 as a, b from range(0, 10000000) tbl(b) order by b desc;
+
+statement ok
+select quantile(a, 0.5) over (
```
Can we check the result here?

Mytherin commented on test/sql/window/test_mode_window.test:
```diff
@@ -0,0 +1,63 @@
+9	3	2
+
+query III
+SELECT r, r/3, mode(r/3) over (order by r rows between 1 preceding and 3 following) FROM modes ORDER BY 1, 2
```
* Could we add a test with a bunch of scattered null values? e.g. you would order by column a, perform the quantile on column b, and column b is basically 50/50 null/not null.
* Perhaps a test with UNBOUNDED PRECEDING and a test with UNBOUNDED FOLLOWING? Perhaps also a test with both?

hawkfish replied:
This is `MODE` but yes we should hit them both.

hawkfish replied:
These tests also drove a small bug in `Finalize` for `MODE` when the hash table is empty.

Mytherin reviewed (commented):
Thanks for the PR! The code looks good to me. Some more suggestions/ideas for the tests:

Commit:
Issue #1746: Moving QUANTILE

Add more tests for scattered NULLs.

Fix corner case in MODE Finalize
when the hash table is present but empty.

Commit:
Merge branch 'master' into hawkfish-moving-quantile

Mytherin commented:
Thanks, looks great now!
